experimental/Schemes/WeilDivisor.jl

```diff
@@ -0,0 +1,306 @@
+  return D
+end
+
+function WeilDivisor(I::IdealSheaf)
+  D = WeilDivisor(scheme(I), ZZ)
+  D[I] = one(ZZ)
+  return D
+end
+
+### copy constructor
+# Only copies the coefficient dictionary; everything else has parent character.
+function WeilDivisor(D::WeilDivisor) 
+  return WeilDivisor(scheme(D), coefficient_ring(D), copy(coefficient_dict(D)))
+end
+
+function Base.show(io::IO, D::WeilDivisor)
+  if has_name(D)
+    print(io, name_of(D))
+    return
+  end
+  if length(components(D)) == 0
+    println(io, "the zero Weil divisor on $(scheme(D))")
+    return
+  end
+  println(io, "Weil divisor on $(scheme(D)) given as the formal sum")
+  comp = ["$(D[I]) ⋅ $(I)" for I in components(D)]
+  out_str = comp[1]
+  for c in comp[2:end]
+    out_str = out_str* " + " * c
+  end
+  println(io, out_str)
+end
+
+function +(D::T, E::T) where {T<:WeilDivisor}
+  X = scheme(D)
+  X == scheme(E) || error("divisors do not live on the same scheme")
+  R = coefficient_ring(D)
+  R == coefficient_ring(E) || error("coefficient rings do not coincide")
+  C = WeilDivisor(D)
+  for I in components(E)
+    if haskey(coefficient_dict(C), I)
+      c = C[I] + E[I]
+      if iszero(c) 
+        delete!(coefficient_dict(C), I)
+      else 
+        C[I] = c
+      end
+    else
+      coefficient_dict(C)[I] = E[I]
+    end
+  end
+  return C
+end
+
+function -(D::T) where {T<:WeilDivisor}
+  E = WeilDivisor(D)
+  for I in components(E)
+    E[I] = -E[I]
+  end
+  return E
+end
+
+-(D::T, E::T) where {T<:WeilDivisor} = D + (-E)
+
+function *(a::RingElem, E::WeilDivisor)
+  c = coefficient_ring(E)()
+  D = WeilDivisor(E)
+  parent(a) == coefficient_ring(E) ? (c = a)::elem_type(coefficient_ring(E)) : c = coefficient_ring(E)(a)
+  for I in components(D)
+    D[I] = c*D[I]
+  end
+  return D
+end
+
+*(a::Int, E::WeilDivisor) = coefficient_ring(E)(a)*E
+*(a::Integer, E::WeilDivisor) = coefficient_ring(E)(a)*E
+
++(D::WeilDivisor, I::IdealSheaf) = D + WeilDivisor(I)
+
+function ==(D::WeilDivisor, E::WeilDivisor) 
+  keys(coefficient_dict(D)) == keys(coefficient_dict(E)) || return false
+  for I in keys(coefficient_dict(D))
+    D[I] == E[I] || return false
+  end
+  return true
+end
+
+#function intersection(D::T, E::T) where {T<:WeilDivisor}
+#  X = scheme(D)
+#  X == scheme(E) || error("divisors do not live on the same scheme")
+#  R = coefficient_ring(D)
+#  R == coefficient_ring(E) || error("divisors do not have the same coefficient ring")
+#  # prepare a copy of the divisors
+#  D_copy = WeilDivisor(X, R)
+#  E_copy = WeilDivisor(X, R)
+#  # check whether a common refinement of the covering is necessary
+#  CD = covering(D)
+#  CE = covering(E)
+#  if CD != CE
+#    CC, f, g = common_refinement(X, CD, CE)
+#    D_copy = pullback(f, D)
+#    E_copy = pullback(g, E)
+#  else
+#    D_copy = D
+#    E_copy = E
+#  end
+#  # TODO: Work out the intersection
+#end
+
+@Markdown.doc """
+    is_in_linear_system(f::VarietyFunctionFieldElem, D::WeilDivisor; check::Bool=true)
```

Comment:
Should probably be just `in_linear_system` since it is also `in` and not `is_in`.

Reply:
Agreed.